### ISSUE TITLE
Validate documents without workflow

### ISSUE THREAD

**Issue by czosel** [projectcaluma/caluma#509]
Currently, it is impossible to check if a document has been filled out correctly without configuring a workflow (specifically, without a `CompleteWorkflowFormTask` or `CompleteTaskFormTask`). For scenarios where the form part of caluma is integrated in another application it would be helpful to expose the validation state of a document in a query.

The most straight-forward implementation would probably be a new `isValid` property on the `Document` node, but since computing the validation state can be computationally intense, we might have to consider implementing this as a separate query, along the lines of
````graphql
isDocumentValid(id: ID!): DocumentValidationState

DocumentValidationState(
  isValid: Boolean
)
```

**Comment by open-dynaMIX** [comment]
I absolutely agree that form validation should not depend on a workflow.

**Comment by czosel** [comment]
@open-dynaMIX :+1: any feedback on the design? /cc @sliverc

**Comment by sliverc** [comment]
Hmm I am not so sure about this.

Because with a `isDocumentValid` the frontend is still the one enforcing validation by asking backend whether a document is valid  instead of doing it itself. And there is no guarantee that a saved document is actually valid.

I only see this to be enforced by the backend with a status resp. when there is a submit action of a document. And that's currently done with a workflow. So I have a feeling the workflow is still the best even though not the most convenient method or do you have any other suggestions?

**Comment by czosel** [comment]
You're right, if the assumption is that caluma is the "first class" backend. The feature I suggested is useful for the following case:

* There is another backend system which maintens the workflow state of an instance.
* Caluma is used as "form store"
* -> The other backend system needs to ask caluma about the validation state of a document before allowing specific state transitions

In summary, this would be an "integration feature" for cases where a specific part of caluma (the form) is integrated into an existing application.

**Comment by open-dynaMIX** [comment]
If we decide implementing this, I don't think there is a reason for another query. With an `isValid` property on the document, the validation will only run, if this property is actually queried.

**Comment by sliverc** [comment]
Just thought how you wanna make sure that a document is not edited once its submitted but this is solved the way that the visibility layer calls the other backend right?

I guess it is not a big deal to implement this if it helps but I think from a Caluma perspective this shouldn't be something to be promoted.

For `isValid` property or `validateDocument` mutation the difference would be that `isValid` would be a simple boolean as with `validateDocument` we could return error messages as well.

**Comment by open-dynaMIX** [comment]
> For isValid property or validateDocument mutation the difference would be that isValid would be a simple boolean as with validateDocument we could return error messages as well.

That's absolutely something we need to consider.

**Comment by czosel** [comment]
> Just thought how you wanna make sure that a document is not edited once its submitted but this is solved the way that the visibility layer calls the other backend right?

Correct!

> I guess it is not a big deal to implement this if it helps but I think from a Caluma perspective this shouldn't be something to be promoted.

Agreed.

> For isValid property or validateDocument mutation the difference would be that isValid would be a simple boolean as with validateDocument we could return error messages as well.

Good point :slightly_smiling_face:

**Comment by winged** [comment]
What speaks against a `validateDocument` mutation is that it is not, in fact, a mutation.
However we can still build a toplevel connection field that can return the validation status. (See #519 for WIP PR to solve this)

**Comment by open-dynaMIX** [comment]
Resolved by #519
